File: pydss/general/TShapeObjImpl.py
from __pyjamas__ import (ARGERROR,)
from dss.general.TShapeObj import (TShapeObj,)
from dss.general.impl.DSSObjectImpl import (DSSObjectImpl,)
from dss.shared.impl.MathUtil import (MathUtil,)
from dss.general.TShape import (TShape,)
# from org.apache.commons.lang.mutable.MutableDouble import (MutableDouble,)


class TShapeObjImpl(DSSObjectImpl, TShapeObj):
    lastValueAccessed = None
    numPoints = None
    # number of points in curve
    arrayPropertyIndex = None
    stdDevCalculated = None
    mean = MutableDouble()
    stdDev = MutableDouble()
    interval = None
    # =0.0 then random interval (hr)
    hours = None
    # time values (hr) if interval > 0.0 else nil
    TValues = None
    # temperatures

    def __init__(self, parClass, TShapeName):
        super(TShapeObjImpl, self)(parClass)
        self.setName(TShapeName.toLowerCase())
        self.objType = parClass.getDSSClassType()
        self.lastValueAccessed = 0
        self.numPoints = 0
        self.interval = 1.0
        # hr
        self.hours = None
        self.TValues = None
        self.stdDevCalculated = False
        # calculate on demand
        self.arrayPropertyIndex = -1
        self.initPropertyValues(0)

    def getTemperature(self, *args):
        """Get temperatures at specified time, hr.

        This method returns the temperature for the given hour.
        If no points exist in the curve, the result is 0.0
        If there are fewer points than requested, the curve is simply assumed to repeat
        Thus a daily load curve can suffice for a yearly load curve: You just get the
        same day over and over again.
        The value returned is the nearest to the interval requested. Thus if you request
        hour=12.25 and the interval is 1.0, you will get interval 12.
        ---
        Get temperatures by index.
        """
        _0 = args
        _1 = len(args)
        if _1 == 1:
            if isinstance(_0[0], double):
                hr, = _0
                result = 0.0
                # default return value if no points in curve
                if self.numPoints > 0:
                    # handle exceptional cases
                    if self.numPoints == 1:
                        result = self.TValues[0]
                    elif self.interval > 0.0:
                        index = self.Math.round(hr / self.interval)
                        if index > self.numPoints:
                            index = index % self.numPoints
                            # wrap around using remainder
                        if index == 0:
                            index = self.numPoints
                        result = self.TValues[index]
                    else:
                        # for random interval
                        # Start with previous value accessed under the assumption that most
                        # of the time, this function will be called sequentially

                        # Normalize Hr to max hour in curve to get wraparound
                        if hr > self.hours[self.numPoints]:
                            hr = hr - ((hr / self.hours[self.numPoints]) * self.hours[self.numPoints])
                        if self.hours[self.lastValueAccessed] > hr:
                            self.lastValueAccessed = 1
                            # start over from beginning
                        _0 = True
                        i = self.lastValueAccessed
                        while True:
                            if _0 is True:
                                _0 = False
                            else:
                                i += 1
                            if not (i < self.numPoints):
                                break
                            # TODO Check zero based indexing
                            if self.Math.abs(self.hours[i] - hr) < 1.0000000000000002e-05:
                                # if close to an actual point, just use it
                                result = self.TValues[i]
                                self.lastValueAccessed = i
                                return result
                            elif self.hours[i] > hr:
                                # interpolate for temperature
                                self.lastValueAccessed = i - 1
                                result = self.TValues[self.lastValueAccessed] + (((hr - self.hours[self.lastValueAccessed]) / (self.hours[i] - self.hours[self.lastValueAccessed])) * (self.TValues[i] - self.TValues[self.lastValueAccessed]))
                                return result
                        # if we fall through the loop, just use last value
                        self.lastValueAccessed = self.numPoints - 1
                        result = self.TValues[self.numPoints]
                return result
            else:
                i, = _0
                if i < self.numPoints and i >= 0:
                    self.lastValueAccessed = i
                    return self.TValues[i]
                else:
                    return 0.0
        else:
            raise ARGERROR(1, 1)

    def calcMeanandStdDev(self):
        if self.numPoints > 0:
            if self.interval > 0.0:
                MathUtil.RCDMeanandStdDev(self.TValues, self.numPoints, self.mean, self.stdDev)
            else:
                MathUtil.curveMeanAndStdDev(self.TValues, self.hours, self.numPoints, self.mean, self.stdDev)
        self.setPropertyValue(4, String.format.format('%.8g', self.mean.doubleValue()))
        self.setPropertyValue(5, String.format.format('%.8g', self.stdDev.doubleValue()))
        self.stdDevCalculated = True

    def getMean(self):
        if not self.stdDevCalculated:
            self.calcMeanandStdDev()
        return self.mean.doubleValue()

    def getStdDev(self):
        if not self.stdDevCalculated:
            self.calcMeanandStdDev()
        return self.stdDev.doubleValue()

    def getHour(self, i):
        """Get hour corresponding to point index."""
        if self.interval == 0:
            if i < self.numPoints and i >= 0:
                self.lastValueAccessed = i
                return self.hours[i]
            else:
                return 0.0
        else:
            self.lastValueAccessed = i
            return self.hours[i] * self.interval

    def dumpProperties(self, f, complete):
        super(TShapeObjImpl, self).dumpProperties(f, complete)
        _0 = True
        i = 0
        while True:
            if _0 is True:
                _0 = False
            else:
                i += 1
            if not (i < self.getParentClass().getNumProperties()):
                break
            _1 = i
            _2 = False
            while True:
                if _1 == 2:
                    _2 = True
                    print '~ ' + self.getParentClass().getPropertyName()[i] + '=(' + self.getPropertyValue(i) + ')'
                    break
                if (_2 is True) or (_1 == 3):
                    _2 = True
                    print '~ ' + self.getParentClass().getPropertyName()[i] + '=(' + self.getPropertyValue(i) + ')'
                    break
                if True:
                    _2 = True
                    print '~ ' + self.getParentClass().getPropertyName()[i] + '=' + self.getPropertyValue(i)
                    break
                break

    def getPropertyValue(self, index):
        _0 = index
        _1 = False
        while True:
            if _0 == 2:
                _1 = True
                result = '['
                break
            if (_1 is True) or (_0 == 3):
                _1 = True
                result = '['
                break
            if True:
                _1 = True
                result = ''
                break
            break
        _2 = index
        _3 = False
        while True:
            if _2 == 1:
                _3 = True
                result = String.format.format('%.8g', self.interval)
                break
            if (_3 is True) or (_2 == 2):
                _3 = True
                _4 = True
                i = 0
                while True:
                    if _4 is True:
                        _4 = False
                    else:
                        i += 1
                    if not (i < self.numPoints):
                        break
                    result = result + String.format.format('%-g, ', self.TValues[i])
                break
            if (_3 is True) or (_2 == 3):
                _3 = True
                if self.hours is not None:
                    _5 = True
                    i = 0
                    while True:
                        if _5 is True:
                            _5 = False
                        else:
                            i += 1
                        if not (i < self.numPoints):
                            break
                        result = result + String.format.format('%-g, ', self.hours[i])
                break
            if (_3 is True) or (_2 == 4):
                _3 = True
                result = String.format.format('%.8g', self.mean.doubleValue())
                break
            if (_3 is True) or (_2 == 5):
                _3 = True
                result = String.format.format('%.8g', self.stdDev.doubleValue())
                break
            if (_3 is True) or (_2 == 9):
                _3 = True
                result = String.format.format('%.8g', self.interval * 3600.0)
                break
            if (_3 is True) or (_2 == 10):
                _3 = True
                result = String.format.format('%.8g', self.interval * 60.0)
                break
            if True:
                _3 = True
                result = super(TShapeObjImpl, self).getPropertyValue(index)
                break
            break
        _6 = index
        _7 = False
        while True:
            if _6 == 2:
                _7 = True
                result = result + ']'
                break
            if (_7 is True) or (_6 == 3):
                _7 = True
                result = result + ']'
                break
            break
        return result

    def initPropertyValues(self, arrayOffset):
        # FIXME Private method in OpenDSS
        self.setPropertyValue(0, '0')
        # number of points to expect
        self.setPropertyValue(1, '1')
        # default = 1.0 hr;
        self.setPropertyValue(2, '')
        # vector of multiplier values
        self.setPropertyValue(3, '')
        # vextor of hour values
        self.setPropertyValue(4, '0')
        # set the mean (otherwise computed)
        self.setPropertyValue(5, '0')
        # set the std dev (otherwise computed)
        self.setPropertyValue(6, '')
        # switch input to a csvfile
        self.setPropertyValue(7, '')
        # switch input to a binary file of singles
        self.setPropertyValue(8, '')
        # switch input to a binary file of singles
        self.setPropertyValue(9, '3600')
        # seconds
        self.setPropertyValue(10, '60')
        # minutes
        self.setPropertyValue(11, '')
        # action option
        super(TShapeObjImpl, self).initPropertyValues(TShape.NumPropsThisClass)

    def saveToDblFile(self):
        # FIXME Private method in OpenDSS
        raise self.UnsupportedOperationException()

    def saveToSngFile(self):
        raise self.UnsupportedOperationException()

    def setMean(self, value):
        self.stdDevCalculated = True
        self.mean.setValue(value)

    def setNumPoints(self, num):
        self.setPropertyValue(0, String.valueOf.valueOf(num))
        # update property list variable
        # reset array property values to keep them in proper order in save
        if self.arrayPropertyIndex >= 0:
            self.setPropertyValue(self.arrayPropertyIndex, self.getPropertyValue(self.arrayPropertyIndex))
        self.numPoints = num

    def setStdDev(self, stddev):
        self.stdDevCalculated = True
        self.stdDev.setValue(stddev)

    def getNumPoints(self):
        return 0

    def getInterval(self):
        return self.interval

    def getHours(self):
        return self.hours

    def setHours(self, values):
        self.hours = values

    def getTValues(self):
        return self.TValues

    def setTValues(self, values):
        # FIXME Private members in OpenDSS
        self.TValues = values

    def getLastValueAccessed(self):
        return self.lastValueAccessed

    def setLastValueAccessed(self, lastValue):
        self.lastValueAccessed = lastValue

    def getArrayPropertyIndex(self):
        return self.arrayPropertyIndex

    def setArrayPropertyIndex(self, index):
        self.arrayPropertyIndex = index

    def isStdDevCalculated(self):
        return self.stdDevCalculated

    def setStdDevCalculated(self, calculated):
        self.stdDevCalculated = calculated

    def setInterval(self, value):
        self.interval = value

File: core/jdss/src/main/java/com/epri/dss/general/impl/TShapeObjImpl.java
package com.epri.dss.general.impl;

import java.io.PrintStream;

import org.apache.commons.lang.mutable.MutableDouble;

import com.epri.dss.common.DSSClass;
import com.epri.dss.general.TShape;
import com.epri.dss.general.TShapeObj;
import com.epri.dss.shared.impl.MathUtil;

public class TShapeObjImpl extends DSSObjectImpl implements TShapeObj {

	private int lastValueAccessed,
			numPoints;  // number of points in curve
	private int arrayPropertyIndex;

	private boolean stdDevCalculated;
	private MutableDouble mean = new MutableDouble();
	private MutableDouble stdDev = new MutableDouble();

	protected double interval;   // =0.0 then random interval (hr)
	protected double[] hours;    // time values (hr) if interval > 0.0 else nil
	protected double[] TValues;  // temperatures

	public TShapeObjImpl(DSSClass parClass, String TShapeName) {
		super(parClass);
		setName(TShapeName.toLowerCase());
		objType = parClass.getDSSClassType();

		lastValueAccessed = 0;

		numPoints = 0;
		interval  = 1.0;  // hr
		hours     = null;
		TValues   = null;
		stdDevCalculated = false;  // calculate on demand

		arrayPropertyIndex = -1;

		initPropertyValues(0);
	}

	/**
	 * Get temperatures at specified time, hr.
	 *
	 * This method returns the temperature for the given hour.
	 * If no points exist in the curve, the result is 0.0
	 * If there are fewer points than requested, the curve is simply assumed to repeat
	 * Thus a daily load curve can suffice for a yearly load curve: You just get the
	 * same day over and over again.
	 * The value returned is the nearest to the interval requested. Thus if you request
	 * hour=12.25 and the interval is 1.0, you will get interval 12.
	 */
	public double getTemperature(double hr) {
		int index, i;

		double result = 0.0;  // default return value if no points in curve

		if (numPoints > 0)  // handle exceptional cases
			if (numPoints == 1) {
				result = TValues[0];
			} else {
				if (interval > 0.0) {
					index = (int) Math.round(hr / interval);
					if (index > numPoints)
						index = index % numPoints;  // wrap around using remainder
					if (index == 0)
						index = numPoints;
					result = TValues[index];
				} else {
					// for random interval

					/* Start with previous value accessed under the assumption that most
					 * of the time, this function will be called sequentially
					 */

					/* Normalize hr to max hour in curve to get wraparound */
					if (hr > hours[numPoints]) {
						hr = hr - (int) (hr / hours[numPoints - 1]) * hours[numPoints - 1];
					}

					if (hours[lastValueAccessed] > hr)
						lastValueAccessed = 1;  // start over from beginning
					for (i = lastValueAccessed; i < numPoints; i++) {
						if (Math.abs(hours[i] - hr) < 0.00001) {  // if close to an actual point, just use it
							result = TValues[i];
							lastValueAccessed = i;
							return result;
						} else if (hours[i] > hr) {  // interpolate for temperature
							lastValueAccessed = i - 1;
							result = TValues[lastValueAccessed] +
									(hr - hours[lastValueAccessed]) / (hours[i] - hours[lastValueAccessed]) *
									(TValues[i] - TValues[lastValueAccessed]);
							return result;
						}
					}

					// if we fall through the loop, just use last value
					lastValueAccessed = numPoints - 1;
					result            = TValues[numPoints - 1];
				}
			}

		return result;
	}

	private void calcMeanAndStdDev() {
		if (numPoints > 0)
			if (interval > 0.0) {
				MathUtil.meanandStdDev(TValues, numPoints, mean, stdDev);
			} else {
				MathUtil.curveMeanAndStdDev(TValues, hours, numPoints, mean, stdDev);
			}

		setPropertyValue(4, String.format("%.8g", mean.doubleValue()));
		setPropertyValue(5, String.format("%.8g", stdDev.doubleValue()));

		stdDevCalculated = true;
	}

	public double getMean() {
		if (!stdDevCalculated)
			calcMeanAndStdDev();
		return mean.doubleValue();
	}

	public double getStdDev() {
		if (!stdDevCalculated)
			calcMeanAndStdDev();
		return stdDev.doubleValue();
	}

	/**
	 * Get temperatures by index.
	 */
	public double getTemperature(int i) {
		if (i < numPoints && i >= 0) {
			lastValueAccessed = i;
			return TValues[i];
		} else {
			return 0.0;
		}
	}

	/**
	 * Get hour corresponding to point index.
	 */
	public double getHour(int i) {
		if (interval == 0) {
			if (i < numPoints && i >= 0) {
				lastValueAccessed = i;
				return hours[i];
			} else {
				return 0.0;
			}
		} else {
			lastValueAccessed = i;
			return hours[i] * interval;
		}
	}

	@Override
	public void dumpProperties(PrintStream f, boolean complete) {
		super.dumpProperties(f, complete);

		for (int i = 0; i < getParentClass().getNumProperties(); i++) {
			switch (i) {
			case 2:
				f.println("~ " + getParentClass().getPropertyName()[i] + "=(" + getPropertyValue(i) + ")");
				break;
			case 3:
				f.println("~ " + getParentClass().getPropertyName()[i] + "=(" + getPropertyValue(i) + ")");
				break;
			default:
				f.println("~ " + getParentClass().getPropertyName()[i] + "=" + getPropertyValue(i));
				break;
			}
		}
	}

	@Override
	public String getPropertyValue(int index) {
		String result;
		switch (index) {
		case 2:
			result = "[";
			break;
		case 3:
			result = "[";
			break;
		default:
			result = "";
			break;
		}

		switch (index) {
		case 1:
			result = String.format("%.8g", interval);
			break;
		case 2:
			for (int i = 0; i < numPoints; i++)
				result = result + String.format("%-g, " , TValues[i]);
			break;
		case 3:
			if (hours != null)
				for (int i = 0; i < numPoints; i++)
					result = result + String.format("%-g, ", hours[i]);
			break;
		case 4:
			result = String.format("%.8g", mean.doubleValue());
			break;
		case 5:
			result = String.format("%.8g", stdDev.doubleValue());
			break;
		case 9:
			result = String.format("%.8g", interval * 3600.0);
			break;
		case 10:
			result = String.format("%.8g", interval * 60.0);
			break;
		default:
			result = super.getPropertyValue(index);
			break;
		}

		switch (index) {
		case 2:
			result = result + "]";
			break;
		case 3:
			result = result + "]";
			break;
		}

		return result;
	}

	@Override
	public void initPropertyValues(int arrayOffset) {

		setPropertyValue(0, "0");  // number of points to expect
		setPropertyValue(1, "1");  // default = 1.0 hr;
		setPropertyValue(2, "");   // vector of multiplier values
		setPropertyValue(3, "");   // vextor of hour values
		setPropertyValue(4, "0");  // set the mean (otherwise computed)
		setPropertyValue(5, "0");  // set the std dev (otherwise computed)
		setPropertyValue(6, "");   // switch input to a csvfile
		setPropertyValue(7, "");   // switch input to a binary file of singles
		setPropertyValue(8, "");   // switch input to a binary file of singles
		setPropertyValue(9, "3600");  // seconds
		setPropertyValue(10, "60");   // minutes
		setPropertyValue(11, "");  // action option

		super.initPropertyValues(TShape.NumPropsThisClass);
	}

	// FIXME Private method in OpenDSS
	public void saveToDblFile() {
		throw new UnsupportedOperationException();
	}

	// FIXME Private method in OpenDSS
	public void saveToSngFile() {
		throw new UnsupportedOperationException();
	}

	public void setMean(double value) {
		stdDevCalculated = true;
		mean.setValue(value);
	}

	public void setNumPoints(int num) {
		setPropertyValue(0, String.valueOf(num));  // update property list variable

		// reset array property values to keep them in proper order in save

		if (arrayPropertyIndex >= 0)
			setPropertyValue(arrayPropertyIndex, getPropertyValue(arrayPropertyIndex));

		numPoints = num;
	}

	public void setStdDev(double stddev) {
		stdDevCalculated = true;
		stdDev.setValue(stddev);
	}

	public int getNumPoints() {
		return 0;
	}

	public double getInterval() {
		return interval;
	}

	public double[] getHours() {
		return hours;
	}

	public void setHours(double[] values) {
		hours = values;
	}

	public double[] getTValues() {
		return TValues;
	}

	public void setTValues(double[] values) {
		TValues = values;
	}

	// FIXME Private members in OpenDSS

	public int getLastValueAccessed() {
		return lastValueAccessed;
	}

	public void setLastValueAccessed(int lastValue) {
		lastValueAccessed = lastValue;
	}

	public int getArrayPropertyIndex() {
		return arrayPropertyIndex;
	}

	public void setArrayPropertyIndex(int index) {
		arrayPropertyIndex = index;
	}

	public boolean isStdDevCalculated() {
		return stdDevCalculated;
	}

	public void setStdDevCalculated(boolean calculated) {
		stdDevCalculated = calculated;
	}

	public void setInterval(double value) {
		interval = value;
	}

}
